Added usable prefix field for oLj command

deepakchethan opened the pull request:
For issue #12620

Example output for `oLj` now is:

`{"IO_Plugins":[{"Permissions":"rw_","Name":"ar","Description":"Open ar/lib files [ar|lib]://[file//path]","License":"LGPL3","prefix":"ar://, lib://"},{"Permissions":"rw_","Name":"bfdbg","Description":"BrainFuck Debugger (bfdbg://path/to/file)","License":"LGPL3","prefix":"bfdbg://"}....}`

Please let me know if any more changes are to be done

radare commented on libr/io/plugin.c:
```diff
@@ -158,3 +162,105 @@ R_API int r_io_plugin_write_at(RIODesc *desc, ut64 addr, const ut8 *buf, int len
 	return 0;
 }
+
+R_API char *r_io_plugin_get_uri_prefix (RIOPlugin *plugin) {
```
wtf? why is this necessary? why is this hardcoded?

radare replied:
return type should be const char *

deepakchethan replied:
All the plugin prefixes are hardcoded in r_io_ar_plugin_open() method like
```
static bool r_io_ar_plugin_open(RIO *io, const char *file, bool many) {
	return !strncmp ("ar://", file, 5) || !strncmp ("lib://", file, 6);
}
````
and there are no internal fields that specify the prefix
I made another pr by adding an internal field inside the RIOPlugin and modifying 34 files. Since you told me not to modify internal fields a couple of times. I closed that PR and changed to this.

The PR in question is #13012. 

Thanks for taking the time to review. If both my ways are wrong. Feel free to close this issue with a message pointing the right direction.

radare commented on libr/io/plugin.c:
```diff
@@ -158,3 +162,105 @@ R_API int r_io_plugin_write_at(RIODesc *desc, ut64 addr, const ut8 *buf, int len
+	}
+	else if (!strcmp ("bfdbg", pluginName)) {
+		prefix = "bfdbg://";
+	}
```
else goes in the same line as the }

radare commented:
this information must be provided by the plugin, not hardcoded in any place

> On 7 Feb 2019, at 11:37, Deepak Chethan <notifications@github.com> wrote:
> 
> For issue #12620 <https://github.com/radare/radare2/issues/12620>
> Example output for oLj now is:
> 
> {"IO_Plugins":[{"Permissions":"rw_","Name":"ar","Description":"Open ar/lib files [ar|lib]://[file//path]","License":"LGPL3","prefix":"ar://, lib://"},{"Permissions":"rw_","Name":"bfdbg","Description":"BrainFuck Debugger (bfdbg://path/to/file)","License":"LGPL3","prefix":"bfdbg://"}....}
> 
> Please let me know if any more changes are to be done
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/radare/radare2/pull/13013 <https://github.com/radare/radare2/pull/13013>
> Commit Summary
> 
> Added usable prefix field for oLj command
> File Changes
> 
> M libr/include/r_io.h <https://github.com/radare/radare2/pull/13013/files#diff-0> (1)
> M libr/io/plugin.c <https://github.com/radare/radare2/pull/13013/files#diff-1> (108)
> Patch Links:
> 
> https://github.com/radare/radare2/pull/13013.patch <https://github.com/radare/radare2/pull/13013.patch>
> https://github.com/radare/radare2/pull/13013.diff <https://github.com/radare/radare2/pull/13013.diff>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/13013>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA3-luPk1kvkKVzjLHj-sY2xQRk0kFZ8ks5vLAHigaJpZM4anAtr>.
>

deepakchethan commented:
@radare thanks for nudging me in the right direction.

Since there was no internal field in RIOPlugin about the list of prefixes, I have created one. Initialized the same on all available plugins.

Could you review the code again ?

radare commented on libr/io/p/io_ar.c:
```diff
@@ -94,6 +94,7 @@ RIOPlugin r_io_plugin_ar = {
 	.name = "ar",
 	.desc = "Open ar/lib files [ar|lib]://[file//path]",
 	.license = "LGPL3",
+	.prefixes = "ar://, lib://",
```
what about making this field an actual array of char* instead of a plain string?

deepakchethan replied:
char ** type cannot be directly assigned like `char **a = {"one", "two", NULL}`. One way is using `char name[]` but even that requires using `strcpy` if we are assigning after declaration. 
Also since it has no other use other than directly displaying, is it really required?

xarkes replied:
Sorry I didn't see that comment, ideally the json should yield a list for the available prefixes, so it's actually parsable :)
Something like:
```
{"IO_Plugins":[{"Permissions":"rw_","Name":"ar","Description":"Open ar/lib files [ar|lib]://[file//path]","License":"LGPL3","prefixex": ["ar://", "lib://"]}....}
```

deepakchethan replied:
Okay xarxes, I will make the change and push them by tomorrow.

radare commented on libr/include/r_io.h:
```diff
@@ -158,6 +158,7 @@ typedef struct r_io_plugin_t {
 	char *author;
 	char *license;
 	void *widget;
+	char *prefixes;
```
thats wrong, this buffer cant be modified, use const char *.. or const char ** to use {"foo","bar",NULL}

deepakchethan replied:
I will change it to a const buffer.

radare commented:
also the :// is called URI handler, not "prefix", just to clarify concepts and find a better name for that field

deepakchethan commented:
> also the :// is called URI handler, not "prefix", just to clarify concepts and find a better name for that field

Okay @radare, thanks for letting me know. So I'll change it to `uri_handlers` then?

xarkes commented on libr/io/p/io_ar.c:
```diff
@@ -94,6 +94,7 @@ RIOPlugin r_io_plugin_ar = {
 	.name = "ar",
 	.desc = "Open ar/lib files [ar|lib]://[file//path]",
 	.license = "LGPL3",
+	.uri_handlers = "ar://, lib://",
```
Why don't you use an actual list rather than a plain `char *`?

radare replied:
const char ** should be fine, but we can also use comma separated values and use r_str_split_list, but then i would not add a space after the comma

radare commented:
Const char* and it should work if u define a reference

> On 8 Feb 2019, at 17:17, Deepak Chethan <notifications@github.com> wrote:
> 
> @deepakchethan commented on this pull request.
> 
> In libr/io/p/io_ar.c:
> 
> > @@ -94,6 +94,7 @@ RIOPlugin r_io_plugin_ar = {
>  	.name = "ar",
>  	.desc = "Open ar/lib files [ar|lib]://[file//path]",
>  	.license = "LGPL3",
> +	.prefixes = "ar://, lib://",
> char ** type cannot be directly assigned like char **a = {"one", "two", NULL}. One way is using char name[] but even that requires using strcpy if we are assigning after declaration.
> Also since it has no other use other than directly displaying, is it really required?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

radare commented:
I prefer just “uris”

> On 8 Feb 2019, at 15:07, Deepak Chethan <notifications@github.com> wrote:
> 
> also the :// is called URI handler, not "prefix", just to clarify concepts and find a better name for that field
> 
> Okay @radare, thanks for letting me know. So I'll change it to uri_handlers then?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

radare commented on libr/include/r_io.h:
```diff
@@ -158,6 +158,7 @@ typedef struct r_io_plugin_t {
 	char *author;
 	char *license;
 	void *widget;
+	const char *uri_handlers;
```
rename to uris

radare commented:
btw for simplicity creating bindings or plugins in other languages (rust, vala, python ,...) having just char * instead of char ** would be easier, so maybe we can just go for the comma separated string, without a space

deepakchethan commented:
Okay @radare, I will make the changes

deepakchethan commented:
I have converted the field to `uris` and removed the spaces in the `uris` field on all plugins. In the json parsing code. I used the split method to split the uris and display them in the form of the array. 

The raw output now is:
`{"IO_Plugins":[{"permissions":"rw_","name":"ar","description":"Open ar/lib files [ar|lib]://[file//path]","license":"LGPL3","uris":["ar://","lib://"]},{"permissions":"rw_","name":"bfdbg","description":"BrainFuck Debugger (bfdbg://path/to/file)","license":"LGPL3","uris":["bfdbg://"]}..`


The parsed output is: 
<img width="546" alt="screen shot 2019-02-10 at 10 54 50 pm" src="https://user-images.githubusercontent.com/20895544/52537038-e1953e00-2d87-11e9-8bbe-5f8bcabf2d2c.png">

Ready for the review @radare

radare commented:
now we can probably remove the uris from the description, and show them in `r2 -L`

radare commented on libr/io/plugin.c:
```diff
@@ -100,28 +100,53 @@ R_API int r_io_plugin_list(RIO *io) {
+			r_list_foreach (plist, piter, uri) {
+				pj_s (pj, uri);
+			}
+			pj_end (pj);
```
r_list_free (plist); in here

radare replied:
free (uris); its leaking here too

radare commented on libr/io/plugin.c:
```diff
@@ -100,28 +100,53 @@ R_API int r_io_plugin_list(RIO *io) {
-	io->cb_printf("]}");
+	pj_end (pj);
+	pj_end (pj);
+	io->cb_printf (pj_string(pj));
```
missing space before (

radare commented on libr/io/plugin.c:
```diff
@@ -100,28 +100,53 @@ R_API int r_io_plugin_list(RIO *io) {
 R_API int r_io_plugin_list_json(RIO *io) {
 	RIOPlugin *plugin;
 	SdbListIter *iter;
+	PJ *pj = pj_new();
```
missing space

radare commented on libr/io/p/io_zip.c:
```diff
@@ -651,6 +651,7 @@ static int r_io_zip_close(RIODesc *fd) {
 RIOPlugin r_io_plugin_zip = {
 	.name = "zip",
 	.desc = "Open zip files [apk|ipa|zip|zipall]://[file//path]",
```
remove uri information in the description and improve the text

deepakchethan commented on libr/io/p/io_http.c:
```diff
@@ -104,7 +104,7 @@ static RIODesc *__open(RIO *io, const char *pathname, int rw, int mode) {
 RIOPlugin r_io_plugin_http = {
 	.name = "http",
-	.desc = "http get (http://rada.re/)",
+	.desc = "Make http get requests",
```
Done radare. Anymore changes?

radare commented on libr/io/plugin.c:
```diff
@@ -100,28 +103,55 @@ R_API int r_io_plugin_list(RIO *io) {
 	int n = 0;
-	io->cb_printf("{\"IO_Plugins\":[");
+	pj_o (pj);
+	pj_k (pj, "IO_Plugins");
```
i dont like this camel snake case. maybe better to just use "io_plugins"

deepakchethan replied:
I changed it.

radare commented:
please rebase

deepakchethan commented:
I have squashed and rebased the commits @radare 

Thanks for helping me help with this issue.
